Read Tyk secret from tyk.conf

sedkis opened the pull request:
### Problem
Currently, many users have been running into circles trying to understand why their Gateway secret is not working. It's because the entrypoint script is overriding the tyk.conf secret with an environment variable.  If the environment variable doesn't exist, the entrypoint script uses the default license.

This flow is causing users problems because it is not intuitive.

### First change
If the entrypoint script doesn't detect an environment variable, it will read the one from the tyk.conf and put it into the environment variable, instead of automatically assigning the default dev license.  This ensures backwards compatibility for users who are using env variables and not the tyk.conf secret

3 cases i can think of to use this docker image:
1)  using an env variable for the secret - will continue to work as normal.  the entrypoint script will see that there is an env variable set and will not overwrite it
2) no env variable, secret not set in tyk.conf - the entrypoint script will set the env variable to the default license, which is the same behaviour as before these changes.

**the third case is the one that is causing users problems and the point of this pull request:**
3) no env variable, secret is set in tyk.conf - the entrypoint script will set the env variable to the secret in the tyk.conf

### Second change
I also got rid of the env variable TYKLISTENPORT because this image was merely exposing the port.  Which actually does nothing unless the user runs the docker container with the "-p" flag and explicitly opens up that port as well.  It is mostly there to tell users that they have to remember to expose the port when running the container.

asoorm commented on Dockerfile:
```diff
@@ -21,6 +19,7 @@ RUN apt-get update \
  && apt-get install -y --no-install-recommends \
             python3-setuptools \
             libpython3.4 \
+ && apt-get install jq \
```
I would avoid adding jq as a dependency just to resolve this.

sedkis replied:
It's only 200 kb but I can see how if we used that argument everytime we would end up with a bloated Docker image.

The alternative is to use some AWK hack which is possible but ugly.

sedkis replied:
I don't know about the already included packages well enough to see if there's something we can use instead to parse JSON

excieve replied:
I'm not against that as long as added size isn't going to add up much.

At the same time, why add it as an additional `apt-get` line instead of appending to the previous one?

excieve commented on entrypoint.sh:
```diff
@@ -2,6 +2,17 @@
 TYKCONF=/opt/tyk-gateway/tyk.conf
 
+# for backwards compatibility if TYKSECRET is empty, then set TYKSECRET to value in tyk.conf
+if [[ -z "${TYKSECRET}" ]]; then
```
Let's imagine I'm using the normal `TYK_GW_SECRET` env variable but not everything else, the logic here would be:

  - `if [[ -z "${TYKSECRET}" ]];` is true so `TYKSECRET` is set from the conf file
  - `if [[ -n "${TYKSECRET}" ]]` is true now so `TYK_GW_SECRET` is set to the value of `TYKSECRET`, which is the config file one

This will break existing setups, including those where setting env vars for configuration is typical, e.g. k8s. We even have such internally and I'm afraid they'll break after this change. Our helm chart also depends on the fact:
https://github.com/TykTechnologies/tyk-helm-chart/blob/master/tyk-pro/templates/deployment-gw-repset.yaml#L62

The idea is that `TYK_GW_SECRET` is the official env var used in containers and packages so it should take precedence over everything else. `TYKSECRET` is deprecated and is only handled here for backwards compat. Any env vars, if set, should override whatever is in the conf file.

sedkis replied:
I see.  I misunderstood which variable was deprecated.  That makes more sense.  Let me fix the flow! Thanks @excieve good eye!

excieve commented on README.md:
```diff
@@ -47,11 +47,10 @@ Documentation for gateway configuration can be found here: https://tyk.io/docs/c
 Alternatively, should you wish to configure tyk using environment variables, then you can find the mappings here: https://tyk.io/docs/configure/environment-variables/
 
-Please note that should you wish to change the the gateway secret or listen port as recommended for a production installation, you will need to set `TYKSECRET` and `TYKLISTENPORT` environment variables. The rest of the environment variables remain as per mappings above.
+Please note that you should set the gateway secret in the `TYKSECRET` environment variable.  If you do not, the entrypoint script will attempt to read the gateway secret into the `TYKSECRET` environment variable from the tyk.conf file.
```
This should document that `TYK_GW_SECRET` is what's to be set.

sedkis replied:
Ahhhh.  I had them backwards. will make the changes.

sedkis commented:
1) I updated the readme to use the new secret `TYK_GW_SECRET` instead of the deprecated `TYKSECRET`.

2) i updated the flow to maintain backwards compatibility as per @excieve comment above.

excieve commented on entrypoint.sh:
```diff
@@ -7,7 +7,18 @@ if [[ -n "${TYKSECRET}" ]]; then
+fi
+
+# If no secret found in tyk.conf, will use default license
 if [[ -z "${TYK_GW_SECRET}" ]]; then
```
Don't think this condition would be reachable now. Potentially, we could just ensure that the default value is set in the default conf file, in which case this shouldn't be needed any more.

sedkis replied:
It is reachable if the value of "secret" is empty or it doesn't exist.  but i don't mind removing it altogether and adding the secret to the tyk.conf

sedkis commented:
@excieve , the default tyk.conf already includes a secret with the default license.

Are we good to merge this?
